confuse/templates.py

```diff
@@ -567,6 +567,49 @@ def value(self, view, template=None):
         return pathlib.Path(value)
 
 
+class Optional(Template):
+    """A template that makes a subtemplate optional.
+
+    If the value is present and not null, it must validate against the
+    subtemplate. However, if the value is null or missing, the template will
+    still validate, returning a default value. If `allow_missing` is False,
+    the template will not allow missing values while still permitting null.
+    """
+
+    def __init__(self, subtemplate, default=None, allow_missing=True):
+        self.subtemplate = as_template(subtemplate)
+        if (getattr(self.subtemplate, 'default', REQUIRED) is not REQUIRED
+                and default is None):
+            # Subtemplate's default value will be used if present and
+            # no default was passed to Optional
+            self.default = self.subtemplate.default
```

Comment:
I'm a little surprised by this logic here—snooping for a `.required` attribute on the subtemplate will certainly work for some templates, but it's not exactly part of the interface "contract" that template objects are supposed to expose. I realize this is a tricky question, but maybe we can get the same behavior in a more robust way by just being careful about how we invoke `self.subtemplate.value()`? That is, we could catch `NotFoundError` and similar around that call to do the fallback dynamically.

Reply:
Thanks for the comments. I'm happy to take a different approach, but I'm having a hard time figuring out how to do it. In that section you commented on, I was trying to allow the behavior where `Optional(10)` is equivalent to `Optional(int, default=10)`. Looking at the code for the other templates, I couldn't see a way to determine a subtemplate's default value other than getting it directly. In `Template`, the only method that could return `self.default` is `get_default_value()`:
https://github.com/beetbox/confuse/blob/046ecb143875920e9162483b08f429d24f24d5fe/confuse/templates.py#L72-L77
However, calling that method for a template that doesn't set self.default (like `MappingTemplate`) would raise an `AttributeError`. Would it be better to call `self.subtemplate.get_default_value()` and catch the potential `AttributeError` or `NotFoundError`, instead of the current approach using `getattr`?

If we don't use `get_default_value` or `getattr`, then I think we'd have to rely on `self.subtemplate.value(view)` as suggested. However, for most templates, I think that will only provide the subtemplate's default if the view's value is missing (ie, `view.first()` raises `NotFoundError`):
https://github.com/beetbox/confuse/blob/046ecb143875920e9162483b08f429d24f24d5fe/confuse/templates.py#L57-L70

With `Optional`, the view's value is allowed to be `None`, so calling `self.subtemplate.value(view)` could raise a `ConfigValueError` when the validation fails. We can catch that error, but then we still don't know what the subtemplate's default value is to be able to return it. As far as I could tell, to get the subtemplate's default when the view's value is `None`, I would need to call `self.subtemplate.value(view)` with a `view` that causes `view.first()` to raise `NotFoundError`. Maybe I could make a mock view for that purpose, but that might cause other problems with templates that have defined `.value()` differently. I'll keep looking at it to see if I can come up with another approach, and I'd welcome any other suggestions, but at least right now I'm not seeing it.

As for the documentation, I can put together some examples for that and add it to this PR.

Reply:
Thanks for the summary. I think that captures it perfectly.

I tried option 1, and I ran into two issues that we could probably work around but that may not be ideal:

- The easy one to deal with is that using `RootView([])` in templates.py causes a circular import. I could get around this, though, by doing the import within the `Optional.value()` method.
- The other issue is that `self.subtemplate(RootView([]))` can have somewhat hard to predict behavior for templates that have not defined `self.default`, like `MappingTemplate`. For example, if subtemplate is `MappingTemplate({'foo': int, 'bar': str})`, then `self.subtemplate(RootView([]))` will raise `NotFoundError` (because the `foo` key is not present and required), and `Optional` would have to supply some default value (ie, `None`). However, if subtemplate is `MappingTemplate({'foo': 10, 'bar': 'a string'})`, then `self.subtemplate(RootView([]))` returns `{'foo': 10, 'bar': 'a string'}` as its default value because all its subtemplates have defaults, and `Optional` would return that instead as the default. My concern is that if someone uses `Optional` with a more complicated `MappingTemplate`, they might not expect whatever is returned by default (ie, a single required value would make the entire subtemplate's default switch to `None`). In my original version of `Optional`, having a `MappingTemplate` as the subtemplate always returned a default of `None`, or an explicitly provided default value. That seemed more predictable to me, but I don't mind implementing Option 1 if you think this behavior with `MappingTemplate` would be OK.

Personally, I prefer Option 2, because it makes the interface more consistent to external users of the template, while still allowing a template to internally handle or ignore defaults however it wants. I still need to test this approach out, but I don't think it would cause problems with existing templates and would make the behavior for `Optional` very clear. 

Another idea I had was to stick to the original "snooping" approach, but try to make it clearer that `Optional` was not following the interface and was accessing an attribute that Templates are not required to define. Do you think this approach, with an explicit try/except, would be any better?:
```
    def __init__(self, subtemplate, default=None, allow_missing=True):
        self.subtemplate = as_template(subtemplate)
        try:
            if default is None and self.subtemplate.default is not REQUIRED:
                # For subtemplates that have their own default value, use that
                # value when no default was passed to Optional
                default = self.subtemplate.default
        except AttributeError:
            pass
        self.default = default
        self.allow_missing = allow_missing
```

Reply:
I tested out "Option 2" and changed `get_default_value` so that it doesn't raise an `AttributeError`. The change passed all tests and doesn't seem to impact existing templates in any way (the only templates that don't set `self.default` also never call `self.get_default_value()`). With this change to `get_default_value`, the `Optional` template could call `self.subtemplate.get_default_value()` and use the output of that as the default, ignoring the `NotFoundError` if no subtemplate default is available. I still prefer this approach to "Option 1" with the `self.subtemplate(RootView([]))` trick. If you are OK with me changing `get_default_value` and promoting it to the "official" template interface, I'll make the change and update the PR.

In terms of implementing the new `get_default_value`, I think the most straight-forward approach is to use `hasattr` to test for `self.default` as part of the existing if clause:
```
    def get_default_value(self, key_name='default'):
        if not hasattr(self, 'default') or self.default is REQUIRED:
            # Missing required value. This is an error.
            raise exceptions.NotFoundError(u"{} not found".format(key_name))
        # Missing value, but not required.
        return self.default
```
Alternatively, I could catch the `AttributeError` when `self.default` is accessed but doesn't exist. Do you have a preference on which approach I use?

I also merged in the recent commits and added some documentation for the `Optional` template. Hopefully that helps explain the behavior I am going for.

Reply:
Wow, that's great! Thanks for looking into this. You raise good points about the trouble that passing an "empty" view would cause, so I think you're right that expanding the template interface to include `get_defualt_value` seems like the right thing.

This implementation of `get_default_value` looks good to me; I don't have a preference between the `hasattr` technique or the `AttributeError` technique.

Thank you so much for your attention to detail on this!! Really great work.

Reply:
I updated `get_default_value` to avoid raising `AttributeError` and added a docstring to make it part of the `Template` interface. Since it could now be used outside of the `Template.value()` method and not just in contexts where one is searching for a missing value, I also changed the embedded comments to be a bit more generic.

I also updated the `Optional` template to use `get_default_value` to get the subtemplate's default if necessary. Thanks again for all the feedback on this template. Please let me know if there are any other changes I should make.